### PR TITLE
Add new world special property (corona ztest) 

### PR DESCRIPTION
**Pull request by gta191977649**
### Introduction
This PR adds the ability to control the corona z-testing in the engine, brings the possibility to create the ps2 classical big sun lensflare effect (when z test is disabled), more detail can be found in [here](https://youtu.be/EnvAASQ2els?t=1287)
show
This patch also present in slientpatch as well! i guess people would execpt it as default in SA.
### Summary
- This effect is `enabled` by default, keeping consistent with the SA PC version
- new propname: `coronaztest` is added into the world function, you can set it on or off like other world properties. this includes `setWorldSpecialPropertyEnabled`, `isWorldSpecialPropertyEnabled`
### Sample Useage:
- Enable: `setWorldSpecialPropertyEnabled("coronaztest",true) `
- Disable: `setWorldSpecialPropertyEnabled("coronaztest",false) `
### Test Resource
- command `/ztest (on/off)` to toogle the effect
[coronaztest.zip](https://github.com/multitheftauto/mtasa-blue/files/7941601/coronaztest.zip)

### CODE REVIEW

**Comment by Einheit-101** [comment]
Woah nice, is it possible to port other features of SkyGFX into SA? Like dual pass Alpha rendering to fix the ugly Alpha channels or grassFixPlacement?

**Comment by gta191977649** [comment]
> Woah nice, is it possible to port other features of SkyGFX into SA? Like dual-pass Alpha rendering to fix the ugly Alpha channels or grassFixPlacement?

yeah, I'm working on that at moment, the point is trying to port it as much as I can without touching much of MTA's client code. But I can't guarantee for all skygfx related features can be accurately & properly ported to MTA due to some rw engine related data needs to obtain.
if you interested, it's in [here](https://github.com/gta191977649/MTASA-SkyGfx). :p

**Comment by gta191977649** [comment]
Ready for review.

File: Client/game_sa/CGameSA.cpp [hunk 1]
```diff
@@ -43,8 +43,10 @@ CGameSA::CGameSA()
     m_bAsyncScriptEnabled = false;
     m_bAsyncScriptForced = false;
     m_bASyncLoadingSuspended = false;
+    m_bCoronaZTest = true;
```
Comment (botder) [inline Client/game_sa/CGameSA.cpp]:
```suggestion
```

File: Client/game_sa/CGameSA.h [hunk 2]
```diff
@@ -511,6 +515,7 @@ class CGameSA : public CGame
     bool         m_bASyncLoadingSuspended;
     int          m_iCheckStatus;
     bool         m_bUnderworldWarp;
+    bool         m_bCoronaZTest;
```
Comment (botder) [inline Client/game_sa/CGameSA.h]:
```suggestion
    bool         m_isCoronaZTestEnabled{true};
```

File: Client/game_sa/CGameSA.h [hunk 3]
```diff
@@ -396,6 +397,9 @@ class CGameSA : public CGame
     void SetVehicleSunGlareEnabled(bool bEnabled);
     bool IsVehicleSunGlareEnabled();
 
+    void SetCoronaZTestEnabled(bool bEnabled);
+    bool IsCoronaZTestEnabled();
```
Comment (botder) [inline Client/game_sa/CGameSA.h]:
```suggestion
    void SetCoronaZTestEnabled(bool isEnabled);
    bool IsCoronaZTestEnabled() const noexcept { return m_isCoronaZTestEnabled; }
```

File: Client/game_sa/CGameSA.cpp [hunk 4]
```diff
@@ -746,6 +759,29 @@ bool CGameSA::IsVehicleSunGlareEnabled()
...
+        MemPut<BYTE>(0x6FB17C + 2, 0x20);
+        m_bCoronaZTest = true;
+    }
+}
+
+bool CGameSA::IsCoronaZTestEnabled()
+{
+    return m_bCoronaZTest;
+}
+
```
Comment (botder) [inline Client/game_sa/CGameSA.cpp]:
```suggestion
void CGameSA::SetCoronaZTestEnabled(bool isEnabled)
{
    if (m_isCoronaZTestEnabled == isEnabled)
        return;

    if (isEnabled)
    {
        // Enable ZTest (PC)
        MemPut<BYTE>(0x6FB17C + 0, 0xFF);
        MemPut<BYTE>(0x6FB17C + 1, 0x51);
        MemPut<BYTE>(0x6FB17C + 2, 0x20);
    }
    else
    {
        // Disable ZTest (PS2)
        MemSet((void*)0x6FB17C, 0x90, 3);
    }
    
    m_isCoronaZTestEnabled = isEnabled;
}

```

File: Client/game_sa/CGameSA.cpp [hunk 5]
```diff
@@ -661,8 +673,9 @@ void CGameSA::ResetCheats()
     SetMoonEasterEggEnabled(false);
     SetExtraAirResistanceEnabled(true);
     SetUnderWorldWarpEnabled(true);
+    SetCoronaZTestEnabled(true);
     CVehicleSA::SetVehiclesSunGlareEnabled(false);
-
+
```
Comment (botder) [inline Client/game_sa/CGameSA.cpp]:
```suggestion
```

**Review by botder — CHANGES_REQUESTED** [review]

File: Client/game_sa/CGameSA.cpp [hunk 6]
```diff
@@ -607,6 +608,10 @@ bool CGameSA::IsCheatEnabled(const char* szCheatName)
     if (!strcmp(szCheatName, PROP_VEHICLE_SUNGLARE))
         return IsVehicleSunGlareEnabled();
 
+    if (!strcmp(szCheatName, PROP_CORONA_ZTEST))
+        return IsCoronaZTestEnabled();
+ 
+
```
Comment (botder) [inline Client/game_sa/CGameSA.cpp:614]:
```suggestion
    if (!strcmp(szCheatName, PROP_CORONA_ZTEST))
        return IsCoronaZTestEnabled();
 
```

File: Client/game_sa/CGameSA.cpp [hunk 7]
```diff
@@ -45,6 +45,7 @@ CGameSA::CGameSA()
     m_bASyncLoadingSuspended = false;
     m_iCheckStatus = 0;
 
+
```
Comment (botder) [inline Client/game_sa/CGameSA.cpp]:
```suggestion
```

File: Client/game_sa/CGameSA.cpp [hunk 8]
```diff
@@ -661,8 +672,8 @@ void CGameSA::ResetCheats()
     SetMoonEasterEggEnabled(false);
     SetExtraAirResistanceEnabled(true);
     SetUnderWorldWarpEnabled(true);
+    SetCoronaZTestEnabled(true);
     CVehicleSA::SetVehiclesSunGlareEnabled(false);
```
Comment (botder) [inline Client/game_sa/CGameSA.cpp:675]:
```suggestion
    CVehicleSA::SetVehiclesSunGlareEnabled(false);

```

**Review by botder — CHANGES_REQUESTED** [review]

**Review by botder — APPROVED** [review]

**Pull request merged**